File: go/tests/autonomous_biz_test.go
// autonomous_biz_test.go
//
// Verifies the CircleAI.AutonomousBiz port (autonomous_biz.go): revenue-loop
// publish/subscribe/read, treasury balance derivation (currency-filtered),
// decision-log append + desc read, and null impls.

package circleai_test

import (
	"context"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestAutonomousBiz_RevenueLoopFanoutAndRead(t *testing.T) {
	var loop circleai.InMemoryRevenueLoop
	var seen []string
	unsub := loop.Subscribe(func(e circleai.RevenueEvent) { seen = append(seen, e.EventID) })
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	loop.Publish(circleai.RevenueEvent{EventID: "e1", Amount: circleai.DecimalFromInt(100), Currency: "ZAR", AtUTC: base})
	unsub()
	loop.Publish(circleai.RevenueEvent{EventID: "e2", Amount: circleai.DecimalFromInt(50), Currency: "ZAR", AtUTC: base.Add(time.Hour)})
	if len(seen) != 1 || seen[0] != "e1" {
		t.Fatalf("subscriber saw %v, want [e1] (unsubscribed before e2)", seen)
	}
	got, _ := loop.Read(context.Background(), base.Add(30*time.Minute))
	if len(got) != 1 || got[0].EventID != "e2" {
		t.Fatalf("read since = %+v", got)
	}
}

func TestAutonomousBiz_TreasuryBalanceCurrencyFiltered(t *testing.T) {
	var loop circleai.InMemoryRevenueLoop
	loop.Publish(circleai.RevenueEvent{EventID: "z1", Amount: circleai.DecimalFromInt(100), Currency: "ZAR"})
	loop.Publish(circleai.RevenueEvent{EventID: "u1", Amount: circleai.DecimalFromInt(999), Currency: "USD"})
	loop.Publish(circleai.RevenueEvent{EventID: "z2", Amount: circleai.DecimalFromInt(50), Currency: "zar"}) // case-insensitive
	tr := circleai.NewInMemoryTreasury(&loop, "ZAR")
	snap, err := tr.GetSnapshot(context.Background())
	if err != nil || !snap.Balance.Equal(circleai.DecimalFromInt(150)) || snap.Currency != "ZAR" {
		t.Fatalf("snapshot = %+v err=%v (want 150 ZAR)", snap, err)
	}
}

func TestAutonomousBiz_DecisionLogDescAndLimit(t *testing.T) {
	var log circleai.InMemoryDecisionLog
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	_ = log.Append(context.Background(), circleai.AutonomousDecision{DecisionID: "d1", AtUTC: base})
	_ = log.Append(context.Background(), circleai.AutonomousDecision{DecisionID: "d3", AtUTC: base.Add(2 * time.Hour)})
	_ = log.Append(context.Background(), circleai.AutonomousDecision{DecisionID: "d2", AtUTC: base.Add(time.Hour)})
	got, err := log.Read(context.Background(), 2)
	if err != nil || len(got) != 2 || got[0].DecisionID != "d3" || got[1].DecisionID != "d2" {
		t.Fatalf("decision read desc+limit = %+v err=%v", got, err)
	}
	if _, err := log.Read(context.Background(), 0); err == nil {
		t.Fatalf("limit<=0 must error")
	}
}

func TestAutonomousBiz_NullImpls(t *testing.T) {
	snap, _ := circleai.NullTreasuryInstance.GetSnapshot(context.Background())
	if !snap.Balance.IsZero() || snap.Currency != "ZAR" {
		t.Fatalf("null treasury = %+v", snap)
	}
	circleai.NullRevenueLoopInstance.Subscribe(func(circleai.RevenueEvent) {})() // no-op unsubscribe
	if items, _ := circleai.NullDecisionLogInstance.Read(context.Background(), 5); len(items) != 0 {
		t.Fatalf("null decision log must be empty")
	}
}

File: typescript/tests/autonomousbiz.test.ts
// autonomousbiz.test.ts
// Verifies the CircleAI.AutonomousBiz port: the fan-out revenue loop with kept
// history, the currency-matched running-balance treasury, the append-only
// decision log, and Null* defaults.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryRevenueLoop,
  InMemoryTreasury,
  InMemoryDecisionLog,
  NullTreasury,
  NullRevenueLoop,
  revenueEvent,
  autonomousDecision,
  type RevenueEvent,
} from "../src/autonomousbiz/index";

const D = (s: string): Date => new Date(s);

describe("InMemoryRevenueLoop", () => {
  it("fans out to subscribers and keeps history filtered by since", async () => {
    const loop = new InMemoryRevenueLoop();
    const seen: RevenueEvent[] = [];
    const sub = loop.subscribe(async (e) => {
      seen.push(e);
    });
    loop.publish(revenueEvent("e1", 10, "ZAR", "app", D("2026-01-01T00:00:00Z")));
    loop.publish(revenueEvent("e2", 20, "ZAR", "app", D("2026-03-01T00:00:00Z")));
    sub.dispose();
    loop.publish(revenueEvent("e3", 5, "ZAR", "app", D("2026-04-01T00:00:00Z")));
    // Give async subscribers a tick.
    await new Promise((r) => setTimeout(r, 5));
    assert.equal(seen.length, 2); // e3 arrived after dispose
    const since = await loop.readAsync(D("2026-02-01T00:00:00Z"));
    assert.deepEqual(
      since.map((e) => e.eventId),
      ["e2", "e3"],
    );
  });

  it("does not let a throwing subscriber break publish", () => {
    const loop = new InMemoryRevenueLoop();
    loop.subscribe(async () => {
      throw new Error("bad");
    });
    assert.doesNotThrow(() => loop.publish(revenueEvent("e", 1, "ZAR", "s", new Date())));
  });
});

describe("InMemoryTreasury", () => {
  it("sums currency-matched events into a running balance", async () => {
    const loop = new InMemoryRevenueLoop();
    const treasury = new InMemoryTreasury(loop, "ZAR");
    loop.publish(revenueEvent("e1", 100, "ZAR", "app", new Date()));
    loop.publish(revenueEvent("e2", 50, "zar", "app", new Date())); // case-insensitive currency
    loop.publish(revenueEvent("e3", 999, "USD", "app", new Date())); // wrong currency, excluded
    const snap = await treasury.getSnapshotAsync();
    assert.equal(snap.balance, 150);
    assert.equal(snap.currency, "ZAR");
  });
});

describe("InMemoryDecisionLog", () => {
  it("appends and reads newest-first with a limit", async () => {
    const log = new InMemoryDecisionLog();
    await log.appendAsync(autonomousDecision("d1", "r", "a", D("2026-01-01T00:00:00Z")));
    await log.appendAsync(autonomousDecision("d2", "r", "a", D("2026-02-01T00:00:00Z")));
    const recent = await log.readAsync(1);
    assert.equal(recent.length, 1);
    assert.equal(recent[0].decisionId, "d2");
    await assert.rejects(() => log.readAsync(0));
  });
});

describe("Null autonomous-biz defaults", () => {
  it("NullTreasury returns a zero balance", async () => {
    assert.equal((await NullTreasury.instance.getSnapshotAsync()).balance, 0);
  });
  it("NullRevenueLoop reads empty", async () => {
    assert.equal((await NullRevenueLoop.instance.readAsync(new Date())).length, 0);
  });
});
